### PR TITLE
lib: multicell_location: Added timeout to API

### PR DESCRIPTION
**Pull request by trantanen**
Timeout is passed to multicell location library and time taken for neighbour cell measurement is been subtracted.
I will change the timeouts in location library library into `int32_t timeout`, which will be in milliseconds and SYS_FOREVER_MS will indicate no timeout.
Combine input parameters for multicell_location_get() into a structure to make it easier to modify the API.

### CODE REVIEW

File: include/net/multicell_location.h
Comment (tokangas) [inline include/net/multicell_location.h:76]: Could be `const`, the parameters should be input only.

File: lib/location/method_cellular.c
Comment (tokangas) [inline lib/location/method_cellular.c:116]: There could be a comment for `+ 5`.

**Review by tokangas — APPROVED** [review]

File: lib/location/method_cellular.c
Comment (trantanen) [reply lib/location/method_cellular.c:116]: This is a leftover from testing so will remove this.

File: include/net/multicell_location.h
Comment (trantanen) [reply include/net/multicell_location.h]: Changed

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [inline subsys/net/lib/rest_client/src/rest_client.c:137]: @jhirsi We need to discuss what to do here because `CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT` becomes irrelevant with this implementation when `rest_client_request_defaults_set` is used.

Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c]: Decided to remove `CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT` and `CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT` and use just `CONFIG_REST_CLIENT_REQUEST_TIMEOUT`.

File: subsys/net/lib/rest_client/Kconfig
Comment (tokangas) [inline subsys/net/lib/rest_client/Kconfig:20]: Probably should also mention that timeouts can be disabled by setting this to 0.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (tokangas) [inline subsys/net/lib/rest_client/src/rest_client.c:140]: Could be removed, the timeout is the same as for SO_SNDTIMEO.

**Review by tokangas — APPROVED** [review]

File: lib/multicell_location/services/nrf_cloud_integration.c
Comment (jayteemo) [inline lib/multicell_location/services/nrf_cloud_integration.c:147]: can you update the comment for the `timeout_ms` param: https://github.com/nrfconnect/sdk-nrf/blob/e08f3cf9a17f58358799af7984eb9f3a0fe1140e/include/net/nrf_cloud_rest.h#L68

Comment (jayteemo) [inline lib/multicell_location/services/nrf_cloud_integration.c:147]: the value of `NRF_CLOUD_REST_TIMEOUT_NONE` should be changed to `0` correct?

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:140]: Yep, and I was supposed to simplify previous if condition too due to latest decision to remove send and recv timeout configs. Will do it now.

File: lib/multicell_location/services/nrf_cloud_integration.c
Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: That probably makes sense. I'm doing the change and let's see if we'll find a problem with it.

Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: Done. We have now three places with this similar comment:
`nrf_cloud_rest_context.timeout_ms`
`rest_client_req_context.timeout_ms`
`REST_CLIENT_REQUEST_TIMEOUT`

Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: I will still need to investigate what zero does within http_client library in Zephyr

Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: `http_client` doesn't initialize its timeout handler when timeout is `K_NO_WAIT` (which is 0). However, API description says zero shouldn't be used:
`The timeout value cannot be 0 as there would be no time to receive the data.`

The API doesn't say what the value should be that it wouldn't start the timeout but `SYS_FOREVER_MS` and `0` will not start the timeout handler.
Right now I'm kind of leaning towards keeping the implementation as it is now. The other option would be in my opinion that we would use `SYS_FOREVER_MS` in several components: rest_client, nrf_cloud_rest, multicell_location and location library. It would be possible also to just change rest_client to pass SYS_FOREVER_MS to http_client when rest_client gets zero timeout but I don't know if it's any better than zero given it's not mentioned in the API.
If timeout is not passed to http_client, behavior is then server dependent on when it resets or closes the TCP connection.

What do you think @jhirsi, @tokangas, @jayteemo?

Comment (jayteemo) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: ah, i missed that you are still passing `timeout_ms` to the http_client.
on first glance, i thought this PR was consolidating the timeouts to only use 'timeout_ms` for the socket and to not use the http_client timeout mechanism.

changing `NRF_CLOUD_REST_TIMEOUT_NONE` to `0` is not necessary if the http_client timeout is still being used.

does the http_client timeout mechanism work correctly?

Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: > does the http_client timeout mechanism work correctly?

Yes, it works fine although the timeout in http_client is only for data receiving (not for sending). The timeout is not set if given timeout is `SYS_FOREVER_MS` or `0`.

File: include/net/multicell_location.h
Comment (jhirsi) [inline include/net/multicell_location.h:54]: it is 32 in rest_client, maybe some alignments within a stack

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (jhirsi) [reply subsys/net/lib/rest_client/src/rest_client.c:153]: If timeout occurs in http_client lib, I think that a null response is returned from callback (errno is not set). But if timeout occurs in sckt, I'd assume that recv() returns error, and in that case errno is set (no callback?). 
In both cases http_client_req() just return byte amount that was sent. But the difference is that, from sckt timeout we can get the errno and that might be good to return from rest_client in case of timeout. Although, we could (also) map a null response as a timeout in rest_client? 

Anyways, Can you have a trial how does it really works and what errno is in case of sckt timeout in rest_client http cb handler? I.e. would it be better to have shorter sckt timeout and http_client timeout as "perälauta"? Or do it both, map errno and null response as a timeout?

File: include/net/multicell_location.h
Comment (trantanen) [reply include/net/multicell_location.h:54]: Based on discussion, I think we decided to leave this as it is. I.e., int32_t (ms) in http_client and rest_client, and uint16_t (s) in multicell_location and location libraries.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:153]: As discussed, I've added check for zero HTTP status code into rest_client and that's mapped to `-ETIMEDOUT`. No errno mapping implemented as NULL response results into positive return value from `http_client_req` and we don't know if errno has been set by something else than http_client.

File: lib/multicell_location/services/nrf_cloud_integration.c
Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c:147]: > changing `NRF_CLOUD_REST_TIMEOUT_NONE` to `0` is not necessary if the http_client timeout is still being used.

I think I'd like to change `NRF_CLOUD_REST_TIMEOUT_NONE` to `0` just to align with rest_client, multicell_location library and location library.

**Review by glarsennordic — CHANGES_REQUESTED** [review]
Sorry for the rather scattered comments. I will try to summarize the main points here so that it all makes sense, but comments will need to be read for details.

There are three major things that I think should be addressed:

1. `location_core.c` introduced a timeout numbering convention that differs from how Zephyr typically handles timeout numbering. This means that timeout values are not consistent across `http_client`, `rest_client` vs. `multicell_location`. I think we should eliminate this numbering convention difference, and bring our API in line with how Zephyr handles things. That should significantly reduce confusion moving forward. The specific difference is that `location_core` interprets `0` as "indefinite timeout" whereas Zephyr (with exception to sockets) interprets `0` as "time out instantly". Both numbering systems interpret `SYS_FOREVER_MS` as indefinite timeout, but location services will behave in some strange ways if it receives that as a timeout. This PR, as written, would make that behavior inconsistency more pronounced.
2. You create two additional paths by which `location_core_event_cb_timeout` may be called. However, neither of these paths will prevent the original path (via the location_core timeout timer) from executing. Rather than add these additional paths, I recommend you just set all service integrations (possibly including the sockets themselves) to have indefinite timeout, and rely on the location_core timeout timer to take care of imposing your timeout. This would result in simpler/easier to follow code (no more dynamically computing timeouts based on system time), and would solve the potential double-call-of-cb_timeout problem.
3. You obscured rest_client NULL response behind a timeout error, which would break our REST JITP implementation, and generally make distinguishing between a timeout and a NULL response difficult. I recommend against doing this.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (glarsennordic) [reply subsys/net/lib/rest_client/src/rest_client.c:319]: I do not think that passing zero as the timeout to http_client is allowed, no -- http_client will interpret this as "timeout immediately" (I confirmed this by looking at the source, not just reading its docs)

`SYS_FOREVER_MS` would be the correct value to pass for indefinite timeout. In my review I suggest that we change the entire location library to match with this standard. AFAICT, only one timer would need to be changed to achieve this

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (glarsennordic) [reply subsys/net/lib/rest_client/src/rest_client.c:153]: I would not map NULL HTTP response as an error code. It's not technically a to-spec response from an HTTP server, but the REST JITP library relies on receiving a NULL response from nRF Cloud. NULL response is not the same as a timeout; With NULL response, we know that our HTTP request was received in full by the TCP endpoint. The TCP endpoint simply decided to close the connection (gracefully) without sending a response (not even an empty HTTP response payload). While this is abnormal for an HTTP transaction, it is not technically a connection error, and definitely distinct from a timeout

Comment (jhirsi) [reply subsys/net/lib/rest_client/src/rest_client.c:153]: A good point. So, JITP expects no response instead of ok/accepted. Yeah, that is abnormal. 
For me it looked like that null response from http_client lib represents: a timeout/no response, recv() fails / socket closed by remote peer.  I.e. I think that you cannot count on that request was really received by remote server without a http response.

But yeah, if changing this will break current rest jitp functionality and we cannot do this in this PR. Maybe eventually it would be better that http_client lib would be changed/addded to have an option by the app to distinguish all those different cases when waiting for a response. But that a topic of another discussion/PR.

**Review by trantanen — COMMENTED** [review]
I will soon post some changes but these haven't been tested well enough so I'll need to continue next week (this Friday is public holiday in Finland).

I'm trying to close some of the duplicate comments so we'd ideally have one comment per topic and discussion related to that in one place. Feel free to also close comments that are not relevant anymore.

File: lib/location/method_cellular.c
Comment (trantanen) [reply lib/location/method_cellular.c:157]: To be more specific on this, when location_core_timeout timer triggers, it can send location event and indicate that timeout occurred but `multicell_location_get()` will remain running in the background. This will take us into a state where new location_request() cannot be sent until `multicell_location_get()` has really finalized. And client cannot really know when this would be.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (glarsennordic) [reply subsys/net/lib/rest_client/src/rest_client.c]: > it doesn't mean that the server has received all the data

You're technically right -- what it actually means is that the server has received and ACK'ed one or more TCP datagrams, and has decided it doesn't want to hear the rest. This may or may not occur before the entire HTTP request payload is sent to the server. But either way, if you see `http_status_code == 0`, it's because the server sent FIN-ACK with no other response. The only reason I know this is that I'm the author of the PR that added `http_status_code == 0` to Zephyr:

https://github.com/zephyrproject-rtos/zephyr/pull/43453/commits/d31d3bf445c85363db76bf7d8a011cb2a91e9241

If I recall correctly, there isn't any other way for this status code to be generated. The socket has to gracefully close with zero response buffer. Anything else (such as a timeout) will not raise `http_status_code == 0`

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (glarsennordic) [reply subsys/net/lib/rest_client/src/rest_client.c:320]: > is comparison to K_FOREVER and K_NO_WAIT (=0) 

I am frankly not sure how I missed that, good point! Sorry about that! I stand completely corrected!

FWIW, I'm still glad you are switching to `int32_t` and `SYS_FOREVER_MS` instead of `uint16_t` and `0`, since I still think that is more in line with Zephyr nomenclature.

File: lib/location/method_cellular.c
Comment (glarsennordic) [reply lib/location/method_cellular.c:152]: As long as the system workqueue thread is at a non-preempt priority level, this should be true, yeah. But if cell measurement finishes at just the right time, it will be a flip of the coin whether this work function resumes (on line 129 at the semaphore), or whether the system workqueue initiates the timeout expiry function. I don't think that's a huge problem, but it'll create potentially inconsistent behavior


If someone changes the system workqueue thread priority to a preempting priority, then I'm pretty sure this routine could get preempted in the middle of manually detecting a timeout, and in that case all bets are off. For instance, the expiry function in location_core could preempt between lines 150 and 152, in which case you'll still get the double-call.

Comment (glarsennordic) [reply lib/location/method_cellular.c:152]: Ah, yeah, I did not see that. You're right. This should work. I still find it a bit hard to follow, as currently written, and it makes me nervous that we have two separate timeout mechanisms on two separate threads, both of which are set up to execute at almost the exact same time. With non-preempting thread priorities it isn't a problem, but that is because the behavior deviates from intuition (while this work function is executing, the other timeout function literally cannot be executed, and vice-versa)

In fact, I am pretty sure that if cell measurement finishes at just the right time, it will be a coin-toss between whether this work function resumes on line 129, or the the location_core timeout work function initiates. That potential for inconsistency makes me nervous.

In general I think it would be cleaner to have only a single time-out mechanism rather than two timeout mechanisms that are running in parallel.

I guess the other option would be to remove the location_core timeout mechanism, and rely on the service integrations to implement the timeout on their own via their respective protocol layers. But I kind of feel like imposing the timeout at the location_core level is the cleaner approach.

File: lib/location/method_cellular.c
Comment (glarsennordic) [reply lib/location/method_cellular.c:157]: Then, you are right, my suggestion would not work. I did not realize that the `http_client` implementation blocks during the request. I assumed the calling thread was yielding while waiting for response.

I ***think*** `http_client` could be modified to be non-blocking. For instance, checking for `EAGAIN` inside `http_wait_data`, and yielding the calling thread for a short time if that occurs, rather than simply failing. Then we could use a non-blocking socket in the REST client.

If we did that, then it would be possible to rely solely on the `location_core` timeout.

Not sure whether pursuing that approach is worth waiting for the required modification to `http_client`.  Additionally, this is getting into fairly low-level socket behavior, though, and I am not an expert, so I may just be outright wrong about whether it is even possible for `http_client` to yield the calling thread rather than block it.

If it really is not possible to use the `location_core` timeout, then I wonder whether we should have it at all. As I said in my other comment, I think it would be cleaner to have only a single timeout mechanism running at any given point in time. If we are already being forced to implement timeout logic individually in each of the service integrations, we might as well rely on it solely, no?

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c
Comment (glarsennordic) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c:158]: I think you updated this, I'm not sure why the comment doesn't show as outdated. Resolving.

File: lib/multicell_location/services/nrf_cloud_integration.c
Comment (trantanen) [reply lib/multicell_location/services/nrf_cloud_integration.c]: I will close this comment as we have changed this back to SYS_FOREVER_MS.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:154]: This discussion also occurs at comment https://github.com/nrfconnect/sdk-nrf/pull/7978#discussion_r905233649 so I'll close this one.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:320]: I'll close this comment as we are moving to `int32_t` and `SYS_FOREVER_MS`.

File: include/net/multicell_location.h
Comment (trantanen) [reply include/net/multicell_location.h]: I'll close this comment as we've changed timeouts to `int32_t`. See more details in comment https://github.com/nrfconnect/sdk-nrf/pull/7978#discussion_r904673363

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c]: > In my eyes, `http_status_code == 0` should ideally uniquely represent a FIN-ACK from the TCP endpoint with no response buffer. There shouldn't be any other way for this to happen. If there is, I think we should draft a PR in Zephyr to make this no longer the case.
> 
> I think for most things, like a TCP socket timeout, the HTTP client lib will simply reply with an error.

While socket send errors cause `http_client_req` to fail with error code, any errors during HTTP receive won't get notified. They are just successful responses with the number of bytes sent. HTTP callback won't be called and you will likely just see NULL response. `http_client` library sets errno but in multithread environment, I'm a bit cautious in using it, and rest_client doesn't use that information. We could check errno in REST client and return errors based on that but I'm not sure if this error would be always because of the `http_client_req` call failing.

So, I guess you can get NULL response if socket receive timeout expires, server sends TCP RST, or there is any other socket error as well as this scenario you mentioned about a non-HTTP response.

> If you wind up pursuing this, can you notify me of the PR? I have a vested interest since I recently worked on the http_client regarding `http_status_code == 0`

I'm not planning any Zephyr PR's at the moment as we'll start the holiday season after this week and will be back in August. And fixes through Zephyr would be time-consuming anyway.

File: lib/location/method_cellular.c
Comment (trantanen) [reply lib/location/method_cellular.c:152]: Yep, let's discuss in the other comment.

File: lib/location/method_cellular.c
Comment (trantanen) [reply lib/location/method_cellular.c:116]: This should be handled by many other discussions.

File: include/net/multicell_location.h
Comment (trantanen) [reply include/net/multicell_location.h]: Handled by other comments and discussions already.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c]: This should be cleaned up now.

File: lib/location/method_cellular.c
Comment (trantanen) [reply lib/location/method_cellular.c:157]: First some comments generally on what the problems are at the moment. You wrote in comment https://github.com/nrfconnect/sdk-nrf/pull/7978#discussion_r905395148:
> In fact, I am pretty sure that if cell measurement finishes at just the right time, it will be a coin-toss between whether this work function resumes on line 129, or the the location_core timeout work function initiates. That potential for inconsistency makes me nervous.

This is true. However, right now without this PR, we have much worse problems. HTTP request can be pending for a long time (seconds and tens of seconds). Location core timeout expires and the API returns but new requests cannot really be made as the lower layers are blocked.
This PR is not fixing everything and there are holes when this might still occur. However, time window for these issues to occur will reduce quite a lot through this PR, which is why I'd like to get this done (this week before our holiday season) even if this would be an intermediate step. If we get updates to `http_client` in the future, we are more than happy to update `rest_client`, `multicell_location` and `location` libraries to take advantage of those changes.

I don't disagree w.r.t two timeouts being complicated. Lower layers are not really our friends at the moment to implement nice timeout mechanism with one timeout so we are compromising somewhere anyway.
Using non-blocking sockets could work but I'm not 100% sure about that either.
In addition, I'd like to have HTTP timeout for the entire operation including the sending part, not just receiving.

I don't know yet if we could just disable location core timeout after ncellmeas has finished and just rely on `http_client` and socket timeouts.

Comment (trantanen) [reply lib/location/method_cellular.c:160]: > I don't know yet if we could just disable location core timeout after ncellmeas has finished and just rely on `http_client` and socket timeouts.

I've done changes to stop core timer after ncellmeas and wifi scanning.

Comment (glarsennordic) [reply lib/location/method_cellular.c:160]: Thanks for hearing out and addressing my concerns. I agree that we are kind of between a rock and a hard place due to how `http_client` works. I'm frankly not 100% sure whether we should choose the rock, or the hard place, but that's ultimately your decision as a code owner, not mine. I'll go ahead and remove my request for changes.

File: lib/location/location_core.c
Comment (glarsennordic) [inline lib/location/location_core.c:523]: Do you mean `(timeout != SYS_FOREVER_MS) && (timeout > 0)` ?

Comment (trantanen) [reply lib/location/location_core.c]: Absolutely

File: lib/location/method_wifi.c
Comment (tokangas) [inline lib/location/method_wifi.c:188]: What if timeout is `SYS_FOREVER_MS`, shouldn't that be handled here?

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:315]: I'm not 100% sure but the reason I've kept it here is that connect has also DNS query and TLS setup so I wanted to protect the code from sending negative timeout to http_client in special cases.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:343]: I haven't protected all of the code from zero timeout and this is intentional although inconsistent as location core timer and rest socket timeouts are not set if zero timeout is given. I don't know if we want to check zero everywhere.

Comment (trantanen) [reply subsys/net/lib/rest_client/src/rest_client.c:343]: Maybe I'll add zero handling here so that the config value can have "disabled" meaning too.

File: subsys/net/lib/rest_client/src/rest_client.c
Comment (tokangas) [reply subsys/net/lib/rest_client/src/rest_client.c:314]: Ok, fair enough.

**Comment by trantanen** [comment]
> LGTM other than possible use of `||` where `&&` was meant

@glarsennordic PR has changes requested by you so could you approve unless you have further comments?

**Review by rlubos — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst
Comment (trantanen) [reply doc/nrf/releases/release-notes-changelog.rst]: Yep, fixed.

**Review by melwee01 — APPROVED** [review]

**Review by glarsennordic — APPROVED** [review]
Changes LGTM, but some of these commits need to be squashed I'm pretty sure

**Comment by trantanen** [comment]
> Changes LGTM, but some of these commits need to be squashed I'm pretty sure

I've done commits on component basis. I guess we don't really have a clear guidance on how this should be done, i.e., which commits can be combined as sometimes I've been asked to split the commits.

**Pull request merged**
